// ambrosus-node-contracts/dist/tasks/base_fee.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _task_base = require('./base/task_base');

var _task_base2 = _interopRequireDefault(_task_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BaseFeeTask = function (_TaskBase) {
  (0, _inherits3.default)(BaseFeeTask, _TaskBase);

  function BaseFeeTask(feesWrapper) {
    (0, _classCallCheck3.default)(this, BaseFeeTask);

    var _this = (0, _possibleConstructorReturn3.default)(this, (BaseFeeTask.__proto__ || Object.getPrototypeOf(BaseFeeTask)).call(this));

    _this.feesWrapper = feesWrapper;
    return _this;
  }

  (0, _createClass3.default)(BaseFeeTask, [{
    key: 'execute',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(_ref2) {
        var _ref3 = (0, _toArray3.default)(_ref2),
            command = _ref3[0],
            options = _ref3.slice(1);

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(command === 'get')) {
                  _context.next = 5;
                  break;
                }

                _context.next = 3;
                return this.getBaseFee();

              case 3:
                _context.next = 12;
                break;

              case 5:
                if (!(command === 'set')) {
                  _context.next = 10;
                  break;
                }

                _context.next = 8;
                return this.setBaseFee(options[0]);

              case 8:
                _context.next = 12;
                break;

              case 10:
                console.error('Unknown sub-command, see yarn task fee help');
                process.exit(1);

              case 12:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute(_x) {
        return _ref.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: 'getBaseFee',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.t0 = console;
                _context2.next = 3;
                return this.feesWrapper.feeForUpload(1);

              case 3:
                _context2.t1 = _context2.sent;

                _context2.t0.log.call(_context2.t0, _context2.t1);

              case 5:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getBaseFee() {
        return _ref4.apply(this, arguments);
      }

      return getBaseFee;
    }()
  }, {
    key: 'setBaseFee',
    value: function () {
      var _ref5 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(newBaseFee) {
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                return _context3.abrupt('return', this.feesWrapper.setBaseUploadFee(newBaseFee));

              case 1:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function setBaseFee(_x2) {
        return _ref5.apply(this, arguments);
      }

      return setBaseFee;
    }()
  }, {
    key: 'help',
    value: function help() {
      return {
        options: '[get], [set feeInWei]',
        description: 'get and set base upload fee'
      };
    }
  }]);
  return BaseFeeTask;
}(_task_base2.default); /*
                        Copyright: Ambrosus Inc.
                        Email: tech@ambrosus.com
                        
                        This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                        
                        This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                        */

exports.default = BaseFeeTask;
