// ambrosus-node-contracts/dist/tasks/node_service.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _task_base = require('./base/task_base');

var _task_base2 = _interopRequireDefault(_task_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NodeServiceTask = function (_TaskBase) {
  (0, _inherits3.default)(NodeServiceTask, _TaskBase);

  function NodeServiceTask(nodeAddress, nodeServiceActions) {
    (0, _classCallCheck3.default)(this, NodeServiceTask);

    var _this = (0, _possibleConstructorReturn3.default)(this, (NodeServiceTask.__proto__ || Object.getPrototypeOf(NodeServiceTask)).call(this));

    _this.nodeServiceActions = nodeServiceActions;
    _this.nodeAddress = nodeAddress;
    return _this;
  }

  (0, _createClass3.default)(NodeServiceTask, [{
    key: 'execute',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(_ref2) {
        var _ref3 = (0, _toArray3.default)(_ref2),
            command = _ref3[0],
            options = _ref3.slice(1);

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(command === 'setUrl')) {
                  _context.next = 5;
                  break;
                }

                _context.next = 3;
                return this.setUrl(options[0]);

              case 3:
                _context.next = 7;
                break;

              case 5:
                console.error('Unknown sub-command, use: yarn task nodeService [setUrl]');
                process.exit(1);

              case 7:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute(_x) {
        return _ref.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: 'printUsage',
    value: function printUsage() {
      console.log('\nUsage: yarn task nodeService [setUrl]');
    }
  }, {
    key: 'setUrl',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(url) {
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (url) {
                  _context2.next = 5;
                  break;
                }

                console.error('Invalid parameter for url');
                this.printUsage();
                process.exit(1);
                return _context2.abrupt('return');

              case 5:
                _context2.next = 7;
                return this.nodeServiceActions.setNodeUrl(this.nodeAddress, url);

              case 7:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function setUrl(_x2) {
        return _ref4.apply(this, arguments);
      }

      return setUrl;
    }()
  }, {
    key: 'help',
    value: function help() {
      return {
        options: '[setUrl]',
        description: 'manages node settings post onboarding'
      };
    }
  }]);
  return NodeServiceTask;
}(_task_base2.default); /*
                        Copyright: Ambrosus Inc.
                        Email: tech@ambrosus.com
                        
                        This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                        
                        This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                        */

exports.default = NodeServiceTask;
