// ambrosus-node-contracts/dist/tasks/ownership.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _task_base = require('./base/task_base');

var _task_base2 = _interopRequireDefault(_task_base);

var _web3_tools = require('../utils/web3_tools');

var _Ownable = require('../contracts/Ownable');

var _Ownable2 = _interopRequireDefault(_Ownable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CheckOwnershipTask = function (_TaskBase) {
  (0, _inherits3.default)(CheckOwnershipTask, _TaskBase);

  function CheckOwnershipTask(web3) {
    (0, _classCallCheck3.default)(this, CheckOwnershipTask);

    var _this = (0, _possibleConstructorReturn3.default)(this, (CheckOwnershipTask.__proto__ || Object.getPrototypeOf(CheckOwnershipTask)).call(this));

    _this.web3 = web3;
    return _this;
  }

  (0, _createClass3.default)(CheckOwnershipTask, [{
    key: 'execute',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(args) {
        var _args, contractAddress;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _args = (0, _slicedToArray3.default)(args, 1), contractAddress = _args[0];

                if (!_web3_tools.utils.isAddress(contractAddress)) {
                  _context.next = 6;
                  break;
                }

                _context.next = 4;
                return this.getOwnership(contractAddress);

              case 4:
                _context.next = 8;
                break;

              case 6:
                console.error('Wrong address, use: yarn task checkOwnership [contract address]');
                process.exit(1);

              case 8:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute(_x) {
        return _ref.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: 'getOwnership',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(contractAddress) {
        var contract;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                contract = (0, _web3_tools.loadContract)(this.web3, _Ownable2.default.abi, contractAddress);
                _context2.t0 = console;
                _context2.next = 4;
                return contract.methods.owner().call();

              case 4:
                _context2.t1 = _context2.sent;

                _context2.t0.log.call(_context2.t0, _context2.t1);

              case 6:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getOwnership(_x2) {
        return _ref2.apply(this, arguments);
      }

      return getOwnership;
    }()
  }, {
    key: 'help',
    value: function help() {
      return {
        description: 'check contract ownership'
      };
    }
  }]);
  return CheckOwnershipTask;
}(_task_base2.default); /*
                        Copyright: Ambrosus Inc.
                        Email: tech@ambrosus.com
                        
                        This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                        
                        This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                        */

exports.default = CheckOwnershipTask;
