// ambrosus-node-contracts/dist/tasks/payouts.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _task_base = require('./base/task_base');

var _task_base2 = _interopRequireDefault(_task_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PayoutsTask = function (_TaskBase) {
  (0, _inherits3.default)(PayoutsTask, _TaskBase);

  function PayoutsTask(web3, sender, payoutsActions) {
    (0, _classCallCheck3.default)(this, PayoutsTask);

    var _this = (0, _possibleConstructorReturn3.default)(this, (PayoutsTask.__proto__ || Object.getPrototypeOf(PayoutsTask)).call(this));

    _this.web3 = web3;
    _this.sender = sender;
    _this.payoutsActions = payoutsActions;
    return _this;
  }

  (0, _createClass3.default)(PayoutsTask, [{
    key: 'execute',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(args) {
        var _args, command;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _args = (0, _slicedToArray3.default)(args, 1), command = _args[0];

                if (!(command === 'total')) {
                  _context.next = 6;
                  break;
                }

                _context.next = 4;
                return this.totalAmount();

              case 4:
                _context.next = 18;
                break;

              case 6:
                if (!(command === 'withdraw')) {
                  _context.next = 11;
                  break;
                }

                _context.next = 9;
                return this.withdraw();

              case 9:
                _context.next = 18;
                break;

              case 11:
                if (!(command === 'period')) {
                  _context.next = 16;
                  break;
                }

                _context.next = 14;
                return this.currentPeriod();

              case 14:
                _context.next = 18;
                break;

              case 16:
                console.error('Unknown sub-command, use: yarn task payout [total/withdraw/period]');
                process.exit(1);

              case 18:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute(_x) {
        return _ref.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: 'totalAmount',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
        var amount;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.prev = 0;
                _context2.next = 3;
                return this.payoutsActions.getTotalAvailablePayout();

              case 3:
                amount = _context2.sent;

                console.log('You can withdraw ' + this.web3.utils.fromWei(amount, 'ether') + ' AMB now!');
                _context2.next = 11;
                break;

              case 7:
                _context2.prev = 7;
                _context2.t0 = _context2['catch'](0);

                console.error(_context2.t0.message);
                process.exit(1);

              case 11:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this, [[0, 7]]);
      }));

      function totalAmount() {
        return _ref2.apply(this, arguments);
      }

      return totalAmount;
    }()
  }, {
    key: 'withdraw',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3() {
        var totalAvailablePayout, payoutInAMB;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.prev = 0;
                _context3.next = 3;
                return this.payoutsActions.getTotalAvailablePayout();

              case 3:
                totalAvailablePayout = _context3.sent;

                if (!(totalAvailablePayout === '0')) {
                  _context3.next = 7;
                  break;
                }

                console.log('Nothing to withdraw :(');
                return _context3.abrupt('return');

              case 7:
                _context3.next = 9;
                return this.payoutsActions.withdraw();

              case 9:
                payoutInAMB = this.web3.utils.fromWei(totalAvailablePayout, 'ether');

                console.log(payoutInAMB + ' AMB have been sent to your account');
                _context3.next = 17;
                break;

              case 13:
                _context3.prev = 13;
                _context3.t0 = _context3['catch'](0);

                console.error(_context3.t0.message);
                process.exit(1);

              case 17:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this, [[0, 13]]);
      }));

      function withdraw() {
        return _ref3.apply(this, arguments);
      }

      return withdraw;
    }()
  }, {
    key: 'currentPeriod',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4() {
        var period;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.prev = 0;
                _context4.next = 3;
                return this.payoutsActions.currentPayoutPeriod();

              case 3:
                period = _context4.sent;

                console.log('Current payout period: ' + period);
                _context4.next = 11;
                break;

              case 7:
                _context4.prev = 7;
                _context4.t0 = _context4['catch'](0);

                console.error(_context4.t0.message);
                process.exit(1);

              case 11:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this, [[0, 7]]);
      }));

      function currentPeriod() {
        return _ref4.apply(this, arguments);
      }

      return currentPeriod;
    }()
  }, {
    key: 'help',
    value: function help() {
      return {
        options: '[total/withdraw/period]',
        description: 'operations with payouts'
      };
    }
  }]);
  return PayoutsTask;
}(_task_base2.default); /*
                        Copyright: Ambrosus Inc.
                        Email: tech@ambrosus.com
                        
                        This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                        
                        This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                        */

exports.default = PayoutsTask;
