// ambrosus-node-contracts/dist/tasks/upload.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _task_base = require('./base/task_base');

var _task_base2 = _interopRequireDefault(_task_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var UploadTask = function (_TaskBase) {
  (0, _inherits3.default)(UploadTask, _TaskBase);

  function UploadTask(uploadActions) {
    (0, _classCallCheck3.default)(this, UploadTask);

    var _this = (0, _possibleConstructorReturn3.default)(this, (UploadTask.__proto__ || Object.getPrototypeOf(UploadTask)).call(this));

    _this.uploadActions = uploadActions;
    return _this;
  }

  (0, _createClass3.default)(UploadTask, [{
    key: 'execute',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(args) {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(args.length !== 2)) {
                  _context.next = 5;
                  break;
                }

                console.error('Invalid parameters.');
                this.printUsage();
                _context.next = 7;
                break;

              case 5:
                _context.next = 7;
                return this.uploadBundle(args[0], args[1]);

              case 7:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute(_x) {
        return _ref.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: 'uploadBundle',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(bundleId, storagePeriods) {
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.uploadActions.uploadBundle(bundleId, storagePeriods);

              case 2:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function uploadBundle(_x2, _x3) {
        return _ref2.apply(this, arguments);
      }

      return uploadBundle;
    }()
  }, {
    key: 'printUsage',
    value: function printUsage() {
      console.log('\nUsage: \nyarn task upload [bundleId] [storagePeriod]');
    }
  }, {
    key: 'help',
    value: function help() {
      return {
        options: '[bundleId] [storagePeriods]',
        description: 'uploads empty bundle (for testing purposes mainly)'
      };
    }
  }]);
  return UploadTask;
}(_task_base2.default); /*
                        Copyright: Ambrosus Inc.
                        Email: tech@ambrosus.com
                        
                        This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                        
                        This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                        */

exports.default = UploadTask;
