// ambrosus-node-contracts/dist/actions/whitelist_actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
Copyright: Ambrosus Inc.
Email: tech@ambrosus.com

This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.

This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
*/

var WhitelistActions = function () {
  function WhitelistActions(kycWhitelistWrapper) {
    (0, _classCallCheck3.default)(this, WhitelistActions);

    this.kycWhitelistWrapper = kycWhitelistWrapper;
  }

  (0, _createClass3.default)(WhitelistActions, [{
    key: 'add',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(address, role, requiredDeposit) {
        var kycWhitelist, owner;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                kycWhitelist = this.kycWhitelistWrapper;
                _context.next = 3;
                return kycWhitelist.getOwner();

              case 3:
                owner = _context.sent;

                if (!(owner !== kycWhitelist.defaultAddress)) {
                  _context.next = 6;
                  break;
                }

                throw new Error('You need to be the owner of the KYC whitelist to perform a add action');

              case 6:
                _context.next = 8;
                return kycWhitelist.isWhitelisted(address);

              case 8:
                if (!_context.sent) {
                  _context.next = 10;
                  break;
                }

                throw new Error('The provided address ' + address + ' is already whitelisted');

              case 10:
                _context.next = 12;
                return kycWhitelist.add(address, role, requiredDeposit);

              case 12:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function add(_x, _x2, _x3) {
        return _ref.apply(this, arguments);
      }

      return add;
    }()
  }, {
    key: 'remove',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(address) {
        var kycWhitelist, owner;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                kycWhitelist = this.kycWhitelistWrapper;
                _context2.next = 3;
                return kycWhitelist.getOwner();

              case 3:
                owner = _context2.sent;

                if (!(owner !== kycWhitelist.defaultAddress)) {
                  _context2.next = 6;
                  break;
                }

                throw new Error('You need to be the owner of the KYC whitelist to perform this action');

              case 6:
                _context2.next = 8;
                return kycWhitelist.isWhitelisted(address);

              case 8:
                if (_context2.sent) {
                  _context2.next = 10;
                  break;
                }

                throw new Error('Address ' + address + ' is not whitelisted');

              case 10:
                _context2.next = 12;
                return kycWhitelist.remove(address);

              case 12:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function remove(_x4) {
        return _ref2.apply(this, arguments);
      }

      return remove;
    }()
  }, {
    key: 'get',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(address) {
        var kycWhitelist;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                kycWhitelist = this.kycWhitelistWrapper;
                _context3.next = 3;
                return kycWhitelist.getRoleAssigned(address);

              case 3:
                _context3.t0 = _context3.sent;
                _context3.next = 6;
                return kycWhitelist.getRequiredDeposit(address);

              case 6:
                _context3.t1 = _context3.sent;
                return _context3.abrupt('return', {
                  role: _context3.t0,
                  requiredDeposit: _context3.t1
                });

              case 8:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function get(_x5) {
        return _ref3.apply(this, arguments);
      }

      return get;
    }()
  }, {
    key: 'transferOwnership',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(newOwnerAddress) {
        var kycWhitelist, owner;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                kycWhitelist = this.kycWhitelistWrapper;
                _context4.next = 3;
                return kycWhitelist.getOwner();

              case 3:
                owner = _context4.sent;

                if (!(owner !== kycWhitelist.defaultAddress)) {
                  _context4.next = 6;
                  break;
                }

                throw new Error('You need to be the owner of the KycWhitelist contract to be able to transfer ownership');

              case 6:
                return _context4.abrupt('return', kycWhitelist.transferOwnership(newOwnerAddress));

              case 7:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function transferOwnership(_x6) {
        return _ref4.apply(this, arguments);
      }

      return transferOwnership;
    }()
  }]);
  return WhitelistActions;
}();

exports.default = WhitelistActions;
