// ambrosus-node-contracts/dist/wrappers/kyc_whitelist_wrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _managed_ownable_contract_wrapper = require('./managed_ownable_contract_wrapper');

var _managed_ownable_contract_wrapper2 = _interopRequireDefault(_managed_ownable_contract_wrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var KycWhitelistWrapper = function (_ManagedOwnableContra) {
  (0, _inherits3.default)(KycWhitelistWrapper, _ManagedOwnableContra);

  function KycWhitelistWrapper() {
    (0, _classCallCheck3.default)(this, KycWhitelistWrapper);
    return (0, _possibleConstructorReturn3.default)(this, (KycWhitelistWrapper.__proto__ || Object.getPrototypeOf(KycWhitelistWrapper)).apply(this, arguments));
  }

  (0, _createClass3.default)(KycWhitelistWrapper, [{
    key: 'add',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(address, role, requiredDeposit) {
        var contract;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.contract();

              case 2:
                contract = _context.sent;
                return _context.abrupt('return', this.processTransaction(contract.methods.add(address, role, requiredDeposit)));

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function add(_x, _x2, _x3) {
        return _ref.apply(this, arguments);
      }

      return add;
    }()
  }, {
    key: 'remove',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(address) {
        var contract;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.contract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt('return', this.processTransaction(contract.methods.remove(address)));

              case 4:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function remove(_x4) {
        return _ref2.apply(this, arguments);
      }

      return remove;
    }()
  }, {
    key: 'isWhitelisted',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(address) {
        var contract;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.contract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt('return', contract.methods.isWhitelisted(address).call());

              case 4:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function isWhitelisted(_x5) {
        return _ref3.apply(this, arguments);
      }

      return isWhitelisted;
    }()
  }, {
    key: 'hasRoleAssigned',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(address, role) {
        var contract;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.contract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt('return', contract.methods.hasRoleAssigned(address, role).call());

              case 4:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function hasRoleAssigned(_x6, _x7) {
        return _ref4.apply(this, arguments);
      }

      return hasRoleAssigned;
    }()
  }, {
    key: 'getRequiredDeposit',
    value: function () {
      var _ref5 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5(address) {
        var contract;
        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this.contract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt('return', contract.methods.getRequiredDeposit(address).call());

              case 4:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getRequiredDeposit(_x8) {
        return _ref5.apply(this, arguments);
      }

      return getRequiredDeposit;
    }()
  }, {
    key: 'getRoleAssigned',
    value: function () {
      var _ref6 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee6(address) {
        var contract;
        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this.contract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt('return', contract.methods.getRoleAssigned(address).call());

              case 4:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getRoleAssigned(_x9) {
        return _ref6.apply(this, arguments);
      }

      return getRoleAssigned;
    }()
  }, {
    key: 'getContractName',
    get: function get() {
      return 'kycWhitelist';
    }
  }]);
  return KycWhitelistWrapper;
}(_managed_ownable_contract_wrapper2.default); /*
                                               Copyright: Ambrosus Inc.
                                               Email: tech@ambrosus.com
                                               
                                               This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                                               
                                               This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                                               */

exports.default = KycWhitelistWrapper;
