// ambrosus-node-contracts/dist/wrappers/sheltering_wrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _managed_contract_wrapper = require('./managed_contract_wrapper');

var _managed_contract_wrapper2 = _interopRequireDefault(_managed_contract_wrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ShelteringWrapper = function (_ManagedContractWrapp) {
  (0, _inherits3.default)(ShelteringWrapper, _ManagedContractWrapp);

  function ShelteringWrapper() {
    (0, _classCallCheck3.default)(this, ShelteringWrapper);
    return (0, _possibleConstructorReturn3.default)(this, (ShelteringWrapper.__proto__ || Object.getPrototypeOf(ShelteringWrapper)).apply(this, arguments));
  }

  (0, _createClass3.default)(ShelteringWrapper, [{
    key: 'isSheltering',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(bundleId) {
        var shelterer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.defaultAddress;
        var contract;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.contract();

              case 2:
                contract = _context.sent;
                return _context.abrupt('return', contract.methods.isSheltering(bundleId, shelterer).call());

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function isSheltering(_x) {
        return _ref.apply(this, arguments);
      }

      return isSheltering;
    }()
  }, {
    key: 'addSheltererReward',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(sheltererId, bundleId, reward) {
        var contract;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.contract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt('return', this.processTransaction(contract.methods.addSheltererReward(sheltererId, bundleId), { value: reward }));

              case 4:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function addSheltererReward(_x3, _x4, _x5) {
        return _ref2.apply(this, arguments);
      }

      return addSheltererReward;
    }()
  }, {
    key: 'shelteringExpirationDate',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(bundleId) {
        var contract;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.contract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt('return', contract.methods.getShelteringExpirationDate(bundleId, this.defaultAddress).call());

              case 4:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function shelteringExpirationDate(_x6) {
        return _ref3.apply(this, arguments);
      }

      return shelteringExpirationDate;
    }()
  }, {
    key: 'bundleStoragePeriods',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(bundleId) {
        var contract;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.contract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt('return', contract.methods.getBundleStoragePeriodsCount(bundleId).call());

              case 4:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function bundleStoragePeriods(_x7) {
        return _ref4.apply(this, arguments);
      }

      return bundleStoragePeriods;
    }()
  }, {
    key: 'getBundleUploadBlockNumber',
    value: function () {
      var _ref5 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5(bundleId) {
        var contract, uploadBlockNumber;
        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this.contract();

              case 2:
                contract = _context5.sent;
                _context5.next = 5;
                return contract.methods.getBundleUploadBlockNumber(bundleId).call();

              case 5:
                uploadBlockNumber = _context5.sent;
                return _context5.abrupt('return', uploadBlockNumber === '0' ? null : parseInt(uploadBlockNumber, 10));

              case 7:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getBundleUploadBlockNumber(_x8) {
        return _ref5.apply(this, arguments);
      }

      return getBundleUploadBlockNumber;
    }()
  }, {
    key: 'getBundleUploader',
    value: function () {
      var _ref6 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee6(bundleId) {
        var contract;
        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this.contract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt('return', contract.methods.getBundleUploader(bundleId).call());

              case 4:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getBundleUploader(_x9) {
        return _ref6.apply(this, arguments);
      }

      return getBundleUploader;
    }()
  }, {
    key: 'getContractName',
    get: function get() {
      return 'sheltering';
    }
  }]);
  return ShelteringWrapper;
}(_managed_contract_wrapper2.default); /*
                                       Copyright: Ambrosus Inc.
                                       Email: tech@ambrosus.com
                                       
                                       This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                                       
                                       This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                                       */

exports.default = ShelteringWrapper;
