// ambrosus-node-contracts/dist/wrappers/transfers_event_emitter_wrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _managed_contract_wrapper = require('./managed_contract_wrapper');

var _managed_contract_wrapper2 = _interopRequireDefault(_managed_contract_wrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TransfersEventEmitterWrapper = function (_ManagedContractWrapp) {
  (0, _inherits3.default)(TransfersEventEmitterWrapper, _ManagedContractWrapp);

  function TransfersEventEmitterWrapper() {
    (0, _classCallCheck3.default)(this, TransfersEventEmitterWrapper);
    return (0, _possibleConstructorReturn3.default)(this, (TransfersEventEmitterWrapper.__proto__ || Object.getPrototypeOf(TransfersEventEmitterWrapper)).apply(this, arguments));
  }

  (0, _createClass3.default)(TransfersEventEmitterWrapper, [{
    key: 'transfers',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(fromBlock, toBlock) {
        var contract;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.contract();

              case 2:
                contract = _context.sent;
                return _context.abrupt('return', contract.getPastEvents('TransferStarted', { fromBlock: fromBlock, toBlock: toBlock }));

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function transfers(_x, _x2) {
        return _ref.apply(this, arguments);
      }

      return transfers;
    }()
  }, {
    key: 'resolvedTransfers',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(fromBlock, toBlock) {
        var contract;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.contract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt('return', contract.getPastEvents('TransferResolved', { fromBlock: fromBlock, toBlock: toBlock }));

              case 4:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function resolvedTransfers(_x3, _x4) {
        return _ref2.apply(this, arguments);
      }

      return resolvedTransfers;
    }()
  }, {
    key: 'cancelledTransfers',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(fromBlock, toBlock) {
        var contract;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.contract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt('return', contract.getPastEvents('TransferCancelled', { fromBlock: fromBlock, toBlock: toBlock }));

              case 4:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function cancelledTransfers(_x5, _x6) {
        return _ref3.apply(this, arguments);
      }

      return cancelledTransfers;
    }()
  }, {
    key: 'getContractName',
    get: function get() {
      return 'transfersEventEmitter';
    }
  }]);
  return TransfersEventEmitterWrapper;
}(_managed_contract_wrapper2.default); /*
                                       Copyright: Ambrosus Inc.
                                       Email: tech@ambrosus.com
                                       
                                       This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
                                       
                                       This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
                                       */

exports.default = TransfersEventEmitterWrapper;
